crypto/box/src/hkdf_box.rs

```diff
@@ -0,0 +1,167 @@
+use crate::traits::{CryptoBox, Error};
+
+use aead::{
+    generic_array::{
+        sequence::Concat,
+        typenum::{IsLessOrEqual, Sum, Unsigned, B1},
+        ArrayLength, GenericArray,
+    },
+    Aead, Error as AeadError, NewAead,
+};
+use core::{convert::TryFrom, marker::PhantomData, ops::Add};
+use digest::{BlockInput, Digest, FixedOutput, Input, Reset};
+use hkdf::Hkdf;
+use keys::{GetBytes, Kex, KexSecret, PublicKey};
+use rand_core::{CryptoRng, RngCore};
+
+/// Represents an implementation of Ristretto-Box using Hkdf<Blake2b> and Aes128Gcm
+///
+/// This structure contains the actual cryptographic primitive details, and
+/// specifies part of the wire format of the "footer" where the ephemeral
+/// public key comes first, and the mac comes second.
+///
+/// Note: If when instantiating this, the IsLessOrEqual bounds fail,
+/// it may indicate that your Kex algorithm does not supply enough entropy to properly
+/// drive the Aead algorithm that you have chosen, and so the combination would be insecure.
+pub struct HkdfBox<KexAlgo, DigestAlgo, AeadAlgo>
+where
+    KexAlgo: Kex,
+    for<'privkey> <KexAlgo as Kex>::Public: From<&'privkey <KexAlgo as Kex>::EphemeralPrivate>,
+    DigestAlgo: Digest + Input + FixedOutput + Default + Clone + BlockInput + Reset,
+    AeadAlgo: Aead + NewAead,
+    AeadAlgo::NonceSize:
+        IsLessOrEqual<<KexAlgo::Secret as KexSecret>::EntropyLowerBound, Output = B1>,
+    AeadAlgo::KeySize:
+        IsLessOrEqual<<KexAlgo::Secret as KexSecret>::EntropyLowerBound, Output = B1>,
```

Comment:
This bound is checking that the construction (the combination of Kex or Aead) is correct: The Aead key size does not require more entropy than could be obtained from the shared secret. You need this to satisfy the assumptions of the security proof of the Aead, in general.

This will prevent bad things like trying to drive Aes256Gcm using shared secret extracted from Ristretto -- you won't be able to achieve a 256 bit security level that way.

Reply:
yeah, you're right

I looked up Aes-256-GCM here: https://github.com/RustCrypto/AEADs/blob/master/aes-gcm/src/lib.rs#L142

I guess that this version has a 12 byte nonce, the same size as the Aes-128-GCM version?

It seems weird that the nonce is not as large as the key, and does not get larger as the security parameter increases. Maybe they are being combined somehow?

RistrettoSecret has very slightly less than 32 bytes of entropy, and AES-256-GCM needs 32 bytes of entropy for the key, but that should be fine.

The only way I think that this bound will fail is if the user picks really weird combinations, and as you said off-line, it should be obvious to them that the chain is only as strong as the weakest link. We can offer some guidance in the README